common/src/main/java/com/microsoft/identity/common/internal/cache/HelloCache.kt

```diff
@@ -0,0 +1,352 @@
+ * 2. Client bumps up protocol version.
+ * 3. The targeted app is updated, uninstalled, reinstalled.
+ * 4. Cache entry is expired.
+ */
+/**
+ * @param context              application context.
+ * @param protocolName         name of the protocol that invokes hello().
+ * @param targetAppPackageName package name of the app that this client will hello() with.
+ * @param components           Platform components.
+ * @param timeoutInMs         Cache entry timeout.
+ */
+open class HelloCache (
+    private val context: Context,
+    private val protocolName: String,
+    private val targetAppPackageName: String,
+    components: IPlatformComponents,
+    private val timeoutInMs: Long = sTimeoutInMs
+) {
+    companion object {
+        private val TAG = HelloCache::class.java.simpleName
+        private const val SHARED_PREFERENCE_NAME = "com.microsoft.common.ipc.hello.cache"
+        private var sIsEnabled = true
+        private val sTimeoutInMs = TimeUnit.HOURS.toMillis(4)
+
+        /**
+         * If set to false, Hello cache will be disabled.
+         * When you're developing protocol change, you might not want the cache to be enabled.
+         *
+         *
+         * For debugging only.
+         */
+        fun setIsEnabled(value: Boolean) {
+            synchronized(HelloCache::class.java) {
+                sIsEnabled = value
+            }
+        }
+    }
+
+    private val fileManager: INameValueStorage<String>
+
+    init {
+        fileManager = components.storageSupplier.getUnencryptedNameValueStore(
+            SHARED_PREFERENCE_NAME, String::class.java
+        )
+    }
+
+    /**
+     * Gets the cached negotiated protocol version.
+     * Returns null
+     * - if there is none.
+     * - if there's error retrieving the value.
+     * - if entry cache entry is expired (entry itself is cleared).
+     * else return [HelloCacheResult] with either successful negotiated protocol version or error.
+     * @param clientMinimumProtocolVersion minimum version of the protocol that the client supports.
+     * @param clientMaximumProtocolVersion maximum version of the protocol that to be advertised by the client.
+     */
+    fun getHelloCacheResult(
+        clientMinimumProtocolVersion: String?,
+        clientMaximumProtocolVersion: String
+    ): HelloCacheResult? {
+        val methodTag = "$TAG:tryGetNegotiatedProtocolVersion"
+        if (!sIsEnabled) {
+            Logger.infoPII(methodTag, "hello cache is not enabled.")
+            return null
+        }
+        val key: String = try {
+            getNegotiatedProtocolVersionCacheKey(
+                clientMinimumProtocolVersion,
+                clientMaximumProtocolVersion
+            )
+        } catch (e: PackageManager.NameNotFoundException) {
+            Logger.error(methodTag, "Failed to retrieve key", e)
+            return null
+        }
+        val negotiationValue = fileManager[key]
+        if (StringUtil.isNullOrEmpty(negotiationValue)) {
```

Comment:
nit: negotiationValue.isNullOrEmpty()

(and you shouldn't have to use the !! next line)